src/gsy_e/models/strategy/heat_pump.py

```diff
@@ -0,0 +1,152 @@
+            min_temperatur_C=min_temperatur_C,
+            max_temperatur_C=max_temperatur_C,
+            external_temperatur_C=external_temperatur_C,
+            tank_volume_l=tank_volume_l,
+            average_consumption=average_consumption,
+            source_type=source_type
+        )
+
+        self.preferred_buying_rate = preferred_buying_rate
+
+    def serialize(self):
+        """Return serialised energy params."""
+        return {**self._energy_params.serialize()}
+
+    @property
+    def state(self) -> HeatPumpState:
+        return self._energy_params.state
+
+    def event_market_cycle(self) -> None:
+        super().event_market_cycle()
+        self._post_order_to_new_market(
+            self.area.spot_market, self.area.spot_market.time_slot)
+
+    def event_tick(self):
+        self._update_open_orders()
+
+    def event_bid_traded(self, *, _market_id: str, bid_trade: Trade) -> None:
+        """TODO: to be tested in the frame of GSYE-426"""
+        market = self.area.spot_market
+        if not market:
+            return
+
+        time_slot = bid_trade.time_slot
+        if bid_trade.buyer_id != self.owner.uuid:
+            return
+
+        self._energy_params.event_traded_energy(market, time_slot)
+        self._energy_params.decrement_posted_energy(market, time_slot)
+
+    def remove_order(self, market: "MarketBase", market_slot: DateTime, order_uuid: str):
+        pass
+
+    def post_order(
+            self, market: "MarketBase", market_slot: DateTime, order_rate: float = None, **kwargs):
+
+        if not order_rate:
+            order_rate = self._order_updaters[market][market_slot].get_energy_rate(
+                self.area.now)
+        order_energy_kWh = self._get_energy_buy_energy(order_rate, market_slot)
+
+        if order_energy_kWh <= FLOATING_POINT_TOLERANCE:
+            return
+
+        market.bid(
+            order_rate * order_energy_kWh, order_energy_kWh,
+            buyer=self.owner.name,
+            original_price=order_rate * order_energy_kWh,
+            buyer_origin=self.owner.name,
+            buyer_origin_id=self.owner.uuid,
+            buyer_id=self.owner.uuid,
+            time_slot=market_slot)
+
+    def remove_open_orders(self, market: "MarketBase", market_slot: DateTime):
+        if self.area.is_market_spot(market.id):
+            bids = [bid
+                    for bid in market.bids.values()
+                    if bid.buyer == self.owner.name]
+        else:
+            bids = [bid
+                    for bid in market.slot_bid_mapping[market_slot]
+                    if bid.buyer == self.owner.name]
+
+        for bid in bids:
+            market.delete_bid(bid)
+
+    def _get_energy_buy_energy(self, buy_rate: float, market_slot: DateTime) -> float:
+        if buy_rate > self.preferred_buying_rate:
+            return self._energy_params.energy_loss_kWh.get(market_slot, 1)
+        return self._energy_params.energy_demand_kWh.get(market_slot, 1)
```

Comment:
I think this method is important to be tested despite its small size, but again this can be deferred for the energy parameters tests.

Reply:
Agreed, I will add a test in the next PR.